tests/python/pants_test/task_test_base.py

```diff
@@ -299,3 +303,116 @@ def assert_console_raises(self, exception, **kwargs):
     """
     with self.assertRaises(exception):
       self.execute_console_task(**kwargs)
+
+
+class DeclarativeTaskTestMixin(object):
+  """Experimental mixin combining target descriptions with a file path dict.
+
+  This class should be mixed in to subclasses of `TaskTestBase`!
+
+  NB: `self.populate_target_dict()` should be called in the `setUp()` method to use the target specs
+  specified in `dist_specs`!
+
+  This mixin also allows specifying tasks to be run before or after the task_type() is executed when
+  calling `self.invoke_tasks()`.
+  """
+
+  @classproperty
+  def dist_specs(cls):
+    """This is an informally-specified nested dict.
+
+    Special keys are 'key' (used to index into `self.target_dict`) and 'filemap' (creates files at
+    the specified relative paths). The rest of the keys are fed into `self.make_target()`. An
+    `OrderedDict` of 2-tuples may be used if targets need to be created in a specific order (e.g. if
+    they have dependencies on each other).
+    """
+    raise NotImplementedError('dist_specs must be implemented!')
+
+  @classproperty
+  def run_before_task_types(cls):
+    """
+    By default, we just use a `BuildLocalPythonDistributions` task. When testing with C/C++ targets,
+    we want to compile and link them as well to get the resulting dist to build, so we add those
+    task types here and execute them beforehand.
+    """
+    return []
+
+  @classproperty
+  def run_after_task_types(cls):
+    """Tasks to run after local dists are built, similar to `run_before_task_types`."""
+    return []
+
+  def populate_target_dict(self):
```

Comment:
This helper method doesn't seem like it is specific to this class. You could probably pass in the `dist_specs` as a map, and return the created targets (which you could then pass to `invoke_tasks`). That would allow the method to move up to `TestBase`, and make it useful in more places 

The rest of the Task manipulation stuff could probably stay here.

Reply:
Done!